src/queues.c

```diff
@@ -555,11 +555,19 @@ gint64 queues_get_next_datachange(gint64 time)
 {
         gint64 wakeup_time = G_MAXINT64;
         gint64 next_second = time + S2US(1) - (time % S2US(1));
+        if(next_second - time < 1000) { // In less than 1ms
+                /* Sometimes, the actual sleep will be a few microseconds
+                   shorter than requested. If we were sleeping until next
+                   second, this new run will be just before the turn of second,
+                   and next_second will be less than 5ms away: in this case,
+                   just sleep 1s more. */
+                next_second += S2US(1);
+        }
```

Comment:
Why this change? It seems that changing n->timestamp to n->start should fix the issue. With this change you would be skipping a turn of second, so the notifications aren't updated correctly.

Reply:
The sleep method used does not guarantee that it will be exact. Sometimes, the sleep will be slightly shorter than expected, and thus the next turn of second will be a few microseconds away only. In this case, that happens once every few dozen sleeps, the next datachange will be reported in less than 1ms away, resulting in a useless call to `run`.

Reply:
Well, if the change is less than 1ms away, dunst will still not update the seconds counter on the notification, so the next run would be required to do that. Unless you can make it so that the seconds counter is updated even when you are a few ms before the change, I would not do this.

Reply:
Fair point. It seems to me that:

1. either we do not care for the occasional useless call to `run`;
1. or the clean solution would be to have two thresholds, eg.:
    * sleep until turn of second + 1ms, which will mean that the vast majority of sleeps will end just after the turn of second (this could be [done here](https://github.com/dunst-project/dunst/blob/836fa0f0289e871f8ba371b5caa6bdbbd99fc59f/src/queues.c#L557));
    * yet update notification age if we are after the turn of second (this is already [done here](https://github.com/dunst-project/dunst/blob/836fa0f0289e871f8ba371b5caa6bdbbd99fc59f/src/notification.c#L676)).

Which option do you prefer? 2 looks cleaner to me, but adds a bit of complexity.

Reply:
I would suggest another option that works as follows:

- The sleep duration should be until the exact turn of second
- If the next turn of second is less than a few ms away, go to the next second (as is done now, but I would make it more than 1 ms, say 10ms)
- When updating the notification, if the next second is less than the same 10 ms away, update to the next second already. That should be as easy as:

```
gint64 t_delta = time_monotonic_now() - ((n->timestamp+10000) - (n->timestamp+10000) % S2US(1));
```

The above code could maybe be simplified, but adding 10 ms to the timestamp should be the same as checking 10ms later.

Reply:
Lgtm. I'll implement that as soon as I get the chance.

Reply:
Thanks!

Reply:
Done in this revision.

Reply:
The commit message says 5ms, but the code says 10ms. But other than that it seems good :+1: